### PR TITLE
Pursuit attribute learner

### PR DESCRIPTION
**Pull request by spigo900**
Related to #984.

### CODE REVIEW

**Comment by spigo900** [comment]
@LichMaster98 This seems to be working now. Some caveats:

1. The candidate attribute extraction code heavily restricts the search space down to only hypotheses that look like `object HAS_PROPERTY property`. (Attributes that cover multiple properties are not included in the space of possible hypotheses.)
2. Despite this it still seems pretty fragile, which makes sense as every object should have at least two "noise" attributes (`animate`/`inanimate` and `biggerThan`/`smallerThan`/`aboutTheSameSizeAsLearner`).
3. I haven't tested it except on some very simple and specific situations. I will see about setting up something more realistic tomorrow.

**Comment by spigo900** [comment]
@LichMaster98 I set up a more realistic experiment and the pursuit attribute learner seems to handle it fine. The experiment runs the pursuit attribute learner (using an object recognizer) on the `m9-attributes` curriculum. The experiment runs without any runtime errors, and the pursuit learner learns four of the ~seven colors (green, white, light brown, dark brown) correctly, so this seems to be working. Results:

- It is still fragile, only achieving about 40% accuracy overall.
- One quirk is that it outputs "the my ground" and "the your ground" because it incorrectly learns that "my" and "your" just mean "inanimate."
- It also learns that red and black mean "inanimate," though the language there is fine after I fixed "the red the ground." (It's now generating "the red ground.")

I"ll look into a hack to get rid of the "the your ground" quirk.

**Comment by spigo900** [comment]
@LichMaster98 @denizbeser This seems to be working as it "should."

- Does this candidate extraction method make sense? Is it trimming things too much?
- Separately, is it a problem that the learner is learning (and lexicalizing) the wrong meanings for "red," "black, "your" and "my"?

File: adam/learner/attributes.py [hunk 1]
```diff
@@ -297,3 +304,192 @@ def _update_hypothesis(
             ),
             trim_after_match=pattern_remove_incomplete_region_or_spatial_path,
         )
+
+
+@attrs
+class PursuitAttributeLearnerNew(  # should it be called new? there is no old one as far as I know
```
Comment (LichMaster98) [inline adam/learner/attributes.py:310]: I'd suggest leaving it called new given the rest of the current learners we are using are called 'new'. Before we release the software if we have time we should probably clean this up.

File: adam/learner/attributes.py [hunk 2]
```diff
@@ -297,3 +304,192 @@ def _update_hypothesis(
...
+        immutableset([object_with_attribute, property]) for property in properties
+    ]
+    return immutableset(
+        [
+            whole_scene_perception_graph.subgraph_by_nodes(
+                candidate_attribute_subgraph_nodes
+            )
+            for candidate_attribute_subgraph_nodes in candidate_attribute_subgraph_node_sets
+        ]
+    )
```
Comment (LichMaster98) [inline adam/learner/attributes.py:495]: This is the section of code we need to be sure to fully disclaim in any writeup which uses this learner. We know that this is a 'hack' to substantially cut down on the search space.

**Review by LichMaster98 — COMMENTED** [review]
I've left a few comments on here mostly for future consideration when writing the write-ups of experiments using this code.

File: adam/learner/attributes.py (hunk 2 above)
Comment (spigo900) [reply adam/learner/attributes.py:495]: @LichMaster98 So just to confirm, do we want to keep this hack then?

Comment (LichMaster98) [reply adam/learner/attributes.py]: Yes @spigo900 otherwise the search space for pursuit could grow too large.

Comment (spigo900) [reply adam/learner/attributes.py]: @LichMaster98 Definitely, just thought I'd confirm. Thanks!

**Review by LichMaster98 — APPROVED** [review]

**Pull request merged**